_resources/port1.0/group/python-1.0.tcl

```diff
@@ -65,7 +65,7 @@ default python.consistent_destroot yes
 
 proc python_get_version {} {
     if {[string match py-* [option name]]} {
-        return [string range [option subport] 2 3]
+        return [string range [lindex [split [option subport] -] 0] 2 end]
```

Comment:
This will presumably interact badly with any existing portfile code that assumes `python.version` is a number. Not necessarily a showstopper, but something to consider.

Also, might be better to write `string range $::subport 2 [string first - $::subport]-1` which avoids converting to a list and back?

Reply:
Good point! I replaced this with two new variables for identifying PyPy:

* `${python.impl}`: Identifies whether we're targeting "PyPy" or "CPython". To make them useful for reporting to the user, they use mixed-caps values.
* `${python.portprefix}`: This replaces the common `port:py${python.version}-rootname` dependency idiom with `port:${python.portprefix}-rootname`.

Reply:
I think it might be better to maintain compatibility with every single dependency declaration than to keep the semantics of `python.version` "pure". I like adding `python.impl`, but how about just using `python.branch` when the language version is desired? Or perhaps adding a variable like `python.langver` for that?

Reply:
> I think it might be better to maintain compatibility with every single dependency declaration than to keep the semantics of `python.version` "pure".

The main downside of that is twofold:

1) `${python.version}` is directly specified in the port file, and for the current approach, that's e.g. `36 37 38 pypy36 pypy37`. Using something like `36 37 38 py36 py37` is somewhat confusing to read, and doesn't obviously refer to _PyPy_ to the same extent.
2) This only happens to work because PyPy starts with `py`. Should we want to add support for IronPython or Jython (if they support Python 3) — or perhaps GraalPython or something entirely different — we'll have to change this again.

Considering that the change is mostly trivial, and you have to edit the port to add PyPy support _anyway_, I don't consider it _that_ much of a downside.

> I like adding `python.impl`, but how about just using `python.branch` when the language version is desired? Or perhaps adding a variable like `python.langver` for that?

I'll have to think about that… But that means editing ports that use `python.level` for checking Python language version. There's also the point that `${python.branch}` is a valid float, and that might make the mistake of comparing them incorrectly much easier.

Reply:
Yes, well, that was before the full effects of the alternative had been illustrated. 🙂

The objection before was having to make changes in ports to stay compatible, and it now looks like the current approach means more of that.